cpp/inc/bond/core/nullable.h

```diff
@@ -482,32 +474,37 @@ class nullable<T, Allocator, false>
 #endif
 
 private:
-    template<typename AllocatorT>
-    void destroy(AllocatorT& alloc)
-    {
 #ifndef BOND_NO_CXX11_ALLOCATOR
-        std::allocator_traits<AllocatorT>::destroy(alloc, _value);
+    typedef typename std::allocator_traits<allocator_type>::
+        template rebind_alloc<value_type> rebind_alloc;
+
+    typedef typename std::allocator_traits<rebind_alloc>::pointer real_pointer;
 #else
-        _value->~T();
+    typedef typename allocator_type::template rebind<value_type>::other rebind_alloc;
+
+    typedef pointer real_pointer;
 #endif
-        alloc.deallocate(_value, 1);
-    }
 
-    void destroy(detail::no_allocator&)
+    void delete_value()
     {
-        delete _value;
+#ifndef BOND_NO_CXX11_ALLOCATOR
+        std::allocator_traits<rebind_alloc>::destroy(_alloc, _value);
+#else
+        _value->~T();
+#endif
+        _alloc.deallocate(_value, 1);
     }
 
 #ifndef BOND_NO_CXX11_RVALUE_REFERENCES
     template<typename AllocatorT, typename Arg1>
-    pointer new_value(AllocatorT& alloc, Arg1&& arg1)
+    real_pointer new_value(AllocatorT& alloc, Arg1&& arg1)
     {
-        T* p = alloc.allocate(1);
+        real_pointer p(alloc.allocate(1));
         try
         {
 #ifndef BOND_NO_CXX11_ALLOCATOR
-            std::allocator_traits<AllocatorT>::construct(alloc, p,
-                std::forward<Arg1>(arg1));
+            std::allocator_traits<AllocatorT>::construct(alloc,
+                boost::addressof(*p), std::forward<Arg1>(arg1));
```

Comment:
Okay. I think I see it now. The `*p` in the argument to `boost::addressof()` might invoke the `operator*` of some smart pointer class. That's what we want: otherwise how would we get a reference to the instance of `value_type` we're interested in out of the smart pointer? But then, we really do want to construct in the storage location for said instance regardless of what `value_type::operator&` does.